File: Python/warmup-1/diff21.py
'''
CodingBat Python Activity "diff21" from warmup-1.
codingbat.com
'''

failures = 0

def diff21(n):
    '''
Given an int n, return the absolute difference between n and 21,
except return double the absolute difference if n is over 21.

Note: https://en.wikipedia.org/wiki/Absolute_difference#:~:text=The%20absolute%20difference%20of%20two,corresponding%20to%20x%20and%20y.

diff21(19) → 2
diff21(10) → 11
diff21(21) → 0
diff21(22) → 2
'''
    pass #Replace pass with your code to implement this function





def catchFailures(testCase, result):
    global failures
    
    try:
        assert eval(testCase) == result, "Test Case Failed: "+testCase+ " → "+str(result)
    except Exception as e:
        failures += 1
        print(e)
            


def test():
    global failures
    
    catchFailures('diff21(19)', 2)
    catchFailures('diff21(10)', 11)
    catchFailures('diff21(21)', 0)
    catchFailures('diff21(22)', 2)
    catchFailures('diff21(25)', 8)
    catchFailures('diff21(30)', 18)
    catchFailures('diff21(0)', 21)
    catchFailures('diff21(1)', 20)
    catchFailures('diff21(2)', 19)
    catchFailures('diff21(-1)', 22)
    catchFailures('diff21(-2)', 23)
    catchFailures('diff21(50)', 58)

    if failures > 0:
        print("\r\n "+ str(failures) + " failed test case(s)   :(")
    else:
        print("Good job -- problem solved!")


test()

File: Java/Warmup1/Diff21.java
/**
 * CodingBat Java Activity "diff21" from Logic1.
 * codingbat.com
 */
public class Diff21
{
	private static int failures = 0;
	private static final String METHOD_NAME = "diff21";

	/**
	 * INSTRUCTIONS:
	 * Given an int n, return the absolute difference between n and 21, except 
	 * return double the absolute difference if n is over 21.
	 * 
	 * diff21(19) -> 2
	 * diff21(10) -> 11
	 * diff21(21) -> 0
	 * 
	 */
	public static int diff21(int n)
	{
		// ADD YOUR CODE HERE
	}

	public static void main(String[] args)
	{
		testCase(19, 2);
		testCase(10, 11);
		testCase(21, 0);
		testCase(22, 2);
		testCase(25, 8);
		testCase(30, 18);
		testCase(0, 21);
		testCase(1, 20);
		testCase(2, 19);
		testCase(-1, 22);
		testCase(-2, 23);
		testCase(50, 58);
		
		System.out.println();
		System.out.println(failures + " failed test cases.");
	}

	private static void testCase(int a, int result)
	{
		String r;
		try
		{
			r = ""+diff21(a);
			if (r.equals(""+result))
			{
				System.out.print("PASS: ");
			}
			else
			{
				System.out.print("FAIL: ");
				failures ++;
			}
		}
		catch (Exception e)
		{
			r = e.getMessage();
			System.out.print("FAIL: ");
			failures ++;
		}
		System.out.println(METHOD_NAME+"("+a+") -> "+r + ", EXPECTED: "+result);
	}

}
